agreement/player.go

```diff
@@ -260,6 +275,64 @@ func (p *player) handleCheckpointEvent(r routerHandle, e checkpointEvent) []acti
 		}}
 }
 
+func (p *player) handleWinningPayloadArrival(payload proposal, ver protocol.ConsensusVersion) {
+	// ignoring validatedAt
+	p.payloadArrivals = append(p.payloadArrivals, payload.receivedAt)
+	p.resizePayloadArrivals(ver)
+}
+
+func (p *player) resizePayloadArrivals(ver protocol.ConsensusVersion) {
+	proto := config.Consensus[ver]
+	if len(p.payloadArrivals) > proto.DynamicFilterPayloadArriavalHistory {
+		p.payloadArrivals = p.payloadArrivals[len(p.payloadArrivals)-proto.DynamicFilterPayloadArriavalHistory:]
+	}
+	for len(p.payloadArrivals) < proto.DynamicFilterPayloadArriavalHistory {
+		p.payloadArrivals = append([]time.Duration{FilterTimeout(0, ver)}, p.payloadArrivals...)
+	}
+}
+
+// calculateFilterTimeout chooses the appropriate filter timeout.
+func (p *player) calculateFilterTimeout(ver protocol.ConsensusVersion, tracer *tracer) time.Duration {
+
+	proto := config.Consensus[ver]
+
+	if !proto.DynamicFilterTimeout || p.Period != 0 {
+		// Either dynamic lambda is disabled, or we're not in period 0 and
+		// therefore, can't use dynamic lambda
+		return FilterTimeout(p.Period, ver)
+	}
+
+	var dynamicDelay time.Duration
+	defaultDelay := FilterTimeout(0, ver)
+	if proto.DynamicFilterPayloadArriavalHistory <= 0 {
+		// we don't keep any history, use the default
+		dynamicDelay = defaultDelay
+	} else if proto.DynamicFilterPayloadArriavalHistory > len(p.payloadArrivals) {
+		// not enough samples, use the default
+		dynamicDelay = defaultDelay
+	} else {
+		sortedArrivals := make([]time.Duration, len(p.payloadArrivals))
+		copy(sortedArrivals[:], p.payloadArrivals[:])
+		sort.Slice(sortedArrivals, func(i, j int) bool { return sortedArrivals[i] < sortedArrivals[j] })
+		dynamicDelay = sortedArrivals[proto.DynamicFilterPayloadArriavalHistory-1]
+	}
+
+	// Make sure the dynamic delay is not too small or too large
+	if dynamicDelay < proto.DynamicFilterTimeoutLowerBound {
+		if tracer != nil {
+			tracer.log.Debugf("Calculated dynamic delay = %v for round %v, period %v. It's too low, setting to %v\n", dynamicDelay, p.Round, p.Period, proto.DynamicFilterTimeoutLowerBound)
+		}
+		dynamicDelay = proto.DynamicFilterTimeoutLowerBound
+	} else if dynamicDelay > defaultDelay {
+		dynamicDelay = defaultDelay
+		tracer.log.Warnf("Calculated dynamic delay = %v for round %v, period %v. It's higher than the default %v\n", dynamicDelay, p.Round, p.Period, defaultDelay)
```

Comment:
you should also check if tracer!=nil and use Debuf like the previous if block?

Reply:
you're right. thanks.